larsim/ElectronDrift/SimDriftElectrons_module.cc

```diff
@@ -390,7 +390,7 @@ namespace detsim {
                   tpcGeo.PlanePitch(0, 1) * fRecipDriftVel[1]);
       }
 
-      const int nIonizedElectrons = fISAlg.CalcIonAndScint(detProp, energyDeposit).numElectrons;
+      const int nIonizedElectrons = energyDeposit.NumElectrons();
```

Comment:
This is likely an issue for MicroBooNE: I don't think there's a separate IonAndScint module that is performed for uboone, and so it's likely that the number of electrons in the energy deposit that is being fed is -1 and thus empty waveforms into wirecell. This may also be the current issue for ICARUS.

The dirty/do no harm fix is to check if num electrons is -1, and if so use ISAlg to calculate them. The clean/leave things better than you found them fix would be to borrow the code from IonAndScint module that allows for different ISAlgs (not just Separate, as is done here), and add an option for 'None' which can then be used to key in on to use the energyDeposit number of electrons directly. If that's done, default option should probably be 'Separate', and SBND will definitely need to then specify the 'None' option in configs.

Reply:
Hi Wes, thanks for your comment.

I don't like either of the two solutions.

The clean option is essentially creating a copy of `IonAndScint_module` inside `SimDriftElectrons_module`, which I personally don't feel is a cleaner solution.

The dirty fix won't work if I check `numElectrons == -1`, the default constructor of [`SimEnergyDeposit`](https://github.com/LArSoft/lardataobj/blob/9cc938ecdc5940b1dd7840bd3a95ab6ae51caae4/lardataobj/Simulation/SimEnergyDeposit.h#L65) sets `numElectrons = 0`. Having 0 electrons is not extraordinary. Thus the dirty fix would require to modify `lardataobj` to set `-1` as default. Perhaps checking a combination of various conditions (checking that electrons and photons are == 0, but other variables aren't) could be a sign that the number of electrons hasn't been calculated already.

I'll attempt to see if I can figure out a dirty solution along these lines.

Reply:
Well I came up with a hacky solution to test `energyDeposit` on `icaruscode` and turns out  `ElectronDrift/SimDriftElectrons_module.cc` is not run on `icaruscode`, at least not on the fcl used on the test `icarus_ci_single_g4_quick_test_icaruscode.fcl` nor on `standard_g4_icarus.fcl`.

I suspect MicroBooNE doesn't run it either since it shares the same name of the process in the TimeTracker table

```
simulate:ionization:larsim/ElectronDrift/ShiftEdepSCE
``` 

By the way that the last MicroBoone test of the G4 stage ran OK:
https://dbweb8.fnal.gov:8443/LarCI/app/ns:uboone/storage/docs/2022/08/08/wirecell_g4_uboone%23oJyVHdX.out

Reply:
To summarise, I don't believe this is the reason for the failed MicroBoone test. It's definitively not the reason for the ICARUS failure, which I addressed in https://github.com/SBNSoftware/icaruscode/pull/429